beacon-chain/rpc/validator/assignments.go

```diff
@@ -4,24 +4,90 @@ import (
 	}
+	return vs.duties(ctx, req)
+}
+
+// StreamDuties returns the duties assigned to a list of validators specified
+// in the request object via a server-side stream. The stream sends out new assignments in case
+// a chain re-org occurred.
+func (vs *Server) StreamDuties(req *ethpb.DutiesRequest, stream ethpb.BeaconNodeValidator_StreamDutiesServer) error {
+	if vs.SyncChecker.Syncing() {
+		return status.Error(codes.Unavailable, "Syncing to latest head, not ready to respond")
+	}
 
+	// We compute duties the very first time the endpoint is called.
+	res, err := vs.duties(stream.Context(), req)
+	if err != nil {
+		return status.Errorf(codes.Internal, "Could not compute validator duties: %v", err)
+	}
+	if err := stream.Send(res); err != nil {
+		return status.Errorf(codes.Internal, "Could not send response over stream: %v", err)
+	}
+
+	// We start a for loop which ticks on every epoch or a chain reorg.
+	stateChannel := make(chan *feed.Event, 1)
+	stateSub := vs.StateNotifier.StateFeed().Subscribe(stateChannel)
+	defer stateSub.Unsubscribe()
+	for {
+		select {
+		// Ticks every epoch to submit assignments to connected validator clients.
+		case <-vs.EpochTicker.C():
+			res, err := vs.duties(stream.Context(), req)
+			if err != nil {
+				return status.Errorf(codes.Internal, "Could not compute validator duties: %v", err)
+			}
+			if err := stream.Send(res); err != nil {
+				return status.Errorf(codes.Internal, "Could not send response over stream: %v", err)
+			}
+		case ev := <-stateChannel:
+			// If a reorg occurred, we recompute duties for the connected validator clients
+			// and send another response over the server stream right away.
+			if ev.Type == statefeed.Reorg {
+				data, ok := ev.Data.(*statefeed.ReorgData)
+				if !ok {
+					return status.Errorf(codes.Internal, "Received incorrect data type over reorg feed: %v", data)
+				}
+				newSlotEpoch := helpers.SlotToEpoch(data.NewSlot)
+				oldSlotEpoch := helpers.SlotToEpoch(data.OldSlot)
+				// We only send out new duties if a reorg across epochs occurred, otherwise
+				// validator shufflings would not have changed as a result of a reorg.
+				if newSlotEpoch >= oldSlotEpoch {
+					continue
+				}
+				res, err := vs.duties(stream.Context(), req)
+				if err != nil {
+					return status.Errorf(codes.Internal, "Could not compute validator duties: %v", err)
+				}
+				if err := stream.Send(res); err != nil {
+					return status.Errorf(codes.Internal, "Could not send response over stream: %v", err)
+				}
+			}
+		case <-stream.Context().Done():
+			return status.Error(codes.Canceled, "Stream context canceled")
+		case <-vs.Ctx.Done():
+			return status.Error(codes.Canceled, "RPC context canceled")
+		}
+	}
+}
+
+// Compute the validator duties from the head state's corresponding epoch
+// for validators public key / indices requested.
+func (vs *Server) duties(ctx context.Context, req *ethpb.DutiesRequest) (*ethpb.DutiesResponse, error) {
```

Comment:
we could possibly send nextCommitteeAssignments over the stream. that way validators will know about their assignments ahead of time

Reply:
I want to do this in a separate PR to not further complicate the logic and line count, but yep we definitely need to compute next epoch assignments